Tom scrape pages

tgj505 opened the pull request:
The now `04_pages` scraping task should now be essentially complete, closing https://github.com/datatoolsrc2023/nlrb_data/issues/12, both with PostgreSQL and SQLite3. All comments have been taken into account.

- removed all `.html` file writing from the task.
- tidied up the cleaning, setup, and post-scrape workflow.
- ensured consistency with prior tasks.

tgj505 committed:
beginning pages and participants tasks

Begun structure for pages and participants tasks.

tgj505 committed:
tasks restructuring

Combined participants and dockets tasks.

tgj505 committed:
Merge branch 'main' into tom_scraping_parsing

Brought tom_scraping_parsing up to date with main after migration to new remote repository.

tgj505 committed:
merging main onto working branch

Added csvs to working branch, ran ingest on 100 rows from one csv.

tgj505 committed:
removing extracted csvs

Removed extracted .csvs in favor of .zip. Also added .csvs to .gitignore

tgj505 committed:
tasks/03_pages work

tgj505 committed:
drafting scraper and data architecture

working on the scraper, figuring out mysql

tgj505 committed:
drafting scraper and parser

tgj505 committed:
drafting parser

tgj505 committed:
task 03_pages drafted, pages scraped

multi-threaded scraper collected case pages, still needs testing, Makefile, testing, psql, refactoring

tgj505 committed:
.gitignore too-large files

page files and tar.gz from the page scrape are too large; simply ignoring for now

tgj505 committed:
.gitignore and requirements editing for merge

tgj505 committed:
draft 03_pages tasks, minor revisions on other tasks

tgj505 committed:
removing data detritus

tgj505 committed:
tidying files

tgj505 committed:
editing .gitignore to resolve merge conflict

tgj505 committed:
merging with main to prepare for pull request

tgj505 committed:
scraping task updated draft

neverett commented on tasks/00_ingest/Makefile:
```diff
@@ -1,6 +1,6 @@
 SHELL := /bin/bash
 
-all: setup pre task post
+all: clean setup pre task post
```
`clean` should be run manually for all tasks, so should probably be removed here. (We should probably add an issue to update each tasks's `setup` script to stop execution if the table it's trying to create already exists and has rows.)

tgj505 replied:
yup, good catch. can't remember why i had put the `clean` back in but had removed in a more recent commit that i've yet to push.

neverett replied:
Looks like `clean` is still in this line, maybe a commit didn't get pushed?

neverett commented on tasks/01_cases/clean.py:
```diff
@@ -6,7 +6,7 @@
     """Drop cases table."""
 
-    query = 'DROP TABLE IF EXISTS cases'
+    query = 'DROP TABLE IF EXISTS cases CASCADE'
```
To keep things from having unexpected consequences, this query should only drop the cases table instead of cascading to dependent tables -- the others (allegations, etc) should be dropped by their own task's `clean` scripts.

tgj505 replied:
if i recall correctly, the `clean` on this task errored out on me because of those other dependencies. will take this out.

neverett commented on tasks/03_pages/scraper.py:
```diff
@@ -0,0 +1,170 @@
+        add_to_pages_table(case_id=case_id, 
+                           case_number=case_number,
+                           error=True,
+                           raw_text=False)
```
I'm wondering if this will insert into the table as expected, since there is a `NOT NULL` constraint on the `raw_text` column. You may want to remove that constraint and set `raw_text = NONE` here so the value is set to NULL in the table.

tgj505 replied:
i've refactored this task to write all errors to `scrape.log` instead of directly to the table.

neverett commented on tasks/03_pages/scraper.py:
```diff
@@ -0,0 +1,170 @@
+        add_to_pages_table(case_id=case_id,
+                               case_number=case_number,
+                               error=True,
+                               raw_text=False)
```
similar comment as above, wondering if this will insert as expected with a `NOT NULL` constraint on that column.

neverett commented on tasks/03_pages/scraper.py:
```diff
@@ -0,0 +1,170 @@
+                               case_number=case_number,
+                               error=True,
+                               raw_text=False)
+        return
```
instead of returning here, you might want to do something like:
```
if soup.article is None:
   <log error>
else:
    page = soup.article.prettify()
    try:
        <add page HTML to pages table>
    except Exception as e:
         <handle exception>
```

neverett commented on tasks/03_pages/scraper.py:
```diff
@@ -0,0 +1,170 @@
+    except Exception as e:
+        print(f'Error: {e}, case: {case_number}')
+        logging.warning(f'{case_id}, {case_number}, write error:{e}')
+        add_to_pages_table(case_id=case_id, 
```
If there's an exception writing to the pages table, there's a good chance that same exception will come up again, so you may not want to write to the pages table here

neverett commented on tasks/03_pages/scraper.py:
```diff
@@ -0,0 +1,170 @@
+                               case_number=case_number,
+                               error=False,
+                               raw_text=page)
+        t2 = time.time()
```
you could move these two lines to an `else` clause, since (I believe) they won't run if there's an exception writing to the pages table anyway.

tgj505 commented on tasks/02_allegations/task.py:
```diff
@@ -39,7 +40,7 @@ def main():
         print(f'Cases with allegations: {n}')
         print(f'Processing allegations...')
-        for row in result:
+        for row in tqdm(result):
```
i was twiddling my thumbs waiting without knowing what was happening so i thought the `tqdm` might be useful!

tgj505 commented on tasks/03_pages/task.py:
```diff
@@ -0,0 +1,56 @@
+
+    # run the scraper. can tune the `max_workers` for multi-threading.
+    # note that each thread/worker opens and closes a connection to the db.
+    try:
```
just a comment on the last point: i'm struggling to understand how to shutdown the executor properly during the keyboard interrupt since the script doesn't exit out after a keyboard interrupt.

tgj505 commented on tasks/03_pages/test_db.py:
```diff
@@ -0,0 +1,28 @@
+            if count == 0:
+                raise Exception(f'Expected {db_config.cases} table '
+                                'to be populated, found 0 records')
+            c.execute(pages_query)
```
i had been running the full `make` each time i restarted the scraper after pausing....but not sure why!

tgj505 committed:
Merge branch 'main' of https://github.com/datatoolsrc2023/nlrb_data into main

tgj505 committed:
merging main into scraping branch

tgj505 committed:
restructuring tasks to match main branch

tgj505 commented on tasks/03_pages/Makefile:
```diff
@@ -0,0 +1,31 @@
+
+clean:
+	# make clean, call manually
+	# Remove empty .html files from scraped pages data
```
i'm refactoring out all the file i/o from this task all together

tgj505 commented on tasks/03_pages/task.py:
```diff
@@ -0,0 +1,56 @@
+
+def main():
+    # use the db's `cases` and `pages` table to get a list of pages to scrape
+    with sql.db_cnx() as cnx:
```
makes a lot of sense!

neverett committed:
quick README updates

neverett committed:
Merge pull request #25 from datatoolsrc2023/nikki-update_readme

Quick README updates

tgj505 committed:
tidying 04_pages task

Reduced redundancy in task, cleaned up scraper, and tested all tasks through 04.

neverett commented on .gitignore:
```diff
@@ -1,8 +1,21 @@
 __pycache__/
+.python-version
 .DS_Store
 .vscode
```
I think you'll need to make this `.vscode/` to ignore the directory.

neverett commented on .vscode/settings.json:
```diff
@@ -0,0 +1,3 @@
+{
```
You'll probably want to remove the `.vscode/` directory from this PR before merging. (I think the easiest way to do that is to copy the directory to a temporary place outside of the `nlrb_data` directory, update `gitignore`, `git rm` it from your branch, then copy it back in.)

neverett commented on sql/postgresql/dockets.sql:
```diff
@@ -0,0 +1,13 @@
+USE `nlrb_data`;
+CREATE TABLE IF NOT EXISTS `dockets` (
+    `id` INT PRIMARY KEY NOT NULL AUTO_INCREMENT,
```
You'll want this to be `id SERIAL PRIMARY KEY NOT NULL` for PostgreSQL

tgj505 replied:
thanks for this catch, did not realize the `dockets.sql` pages were on this branch with old mysql syntax. will clean the sql files

tgj505 replied:
going to remove `dockets.sql` from this PR and deal with it in the parsing branch

neverett commented on sql/postgresql/dockets.sql:
```diff
@@ -0,0 +1,13 @@
+    `id` INT PRIMARY KEY NOT NULL AUTO_INCREMENT,
+    `case_id` INT NOT NULL,
+    `case_number` VARCHAR(16) NOT NULL,
+    `date` DATE NOT NULL,
```
You may want to rename this `d_date` or something similar, since `date` is a reserved word in PostgreSQL. Renaming this column also means you can remove the backticks on all the column names, since you won't be dealing with reserved words or special characters.

neverett commented on sql/postgresql/dockets.sql:
```diff
@@ -0,0 +1,13 @@
+      FOREIGN KEY (case_id) REFERENCES cases (id)
+      ON DELETE CASCADE
+      ON UPDATE CASCADE
+) ENGINE=InnoDB CHARSET=utf8mb4;
```
`ENGINE=Innodb CHARSET=utf8mb4` is only for MySQL, so it's safe to remove that here

neverett commented on tasks/00_ingest/clean.py:
```diff
@@ -6,7 +6,7 @@
     """Drop cases_raw table"""
 
-    query = 'DROP TABLE IF EXISTS cases_raw'
+    query = 'DROP TABLE IF EXISTS cases_raw;'
```
No need to remove this, but just a heads up that the semicolon isn't needed for this sort of query

neverett commented on tasks/04_pages/Makefile:
```diff
@@ -0,0 +1,24 @@
+	which python3
+	python3 ./setup.py
+
+test:
```
Super minor suggestion, but you could call this task `pre` to match the other Makefiles and make it clear this runs before the main task. (The name of the script, `test_db.py` can stay the same.)

neverett commented on tasks/04_pages/clean.py:
```diff
@@ -0,0 +1,25 @@
+
+
+def main():
+    """Drop pages table."""
```
I think you will also need to update the `error_log` table here too -- you can probably do something similar to what's in the `allegations/clean.py` script for that.

neverett commented on tasks/04_pages/test_db.py:
```diff
@@ -0,0 +1,38 @@
+
+    cases_query = 'SELECT COUNT(*) c from cases;'
+    
+    if db_config.db_type == 'sqlite':
```
If the main task has to be restarted, do you expect that folks will rerun `make` or just rerun `task`, then `post`? If the latter, it may be more efficient here to check that the `pages` table has 0 rows instead of checking for existence of the table -- if the table doesn't exist, an exception will be raised, and if this test should be run before table is populated, you'll want to make sure there are 0 rows in it before proceeding.

tgj505 replied:
I wasn't sure which strategy for restarting the task would be more effective. My mental model was to just rerun `make`, but I can see how his might lead to trouble: in cases where a `pages` table had been populated in a prior run of the full project, but then wasn't cleaned or reset when beginning anew. 

A case I was concerned about was when the task had to be restarted, but the sql service hadn't been restarted, which is why I thought it might be worth running the test_db. But `task .py` won't continue if it's unable to connect to the db anyways.

I'll alter this and update the task's `README.md`.

neverett commented on tasks/04_pages/test_db.py:
```diff
@@ -0,0 +1,38 @@
+                raise Exception(f'Expected {db_config.cases} table to be populated, found 0 records')
+            c.execute(pages_query)
+    except Exception as e:
+        raise Exception(f'Issue communicating with db: {e}')
```
I think this won' work as expected, and you'll need to do something like the following instead:
```
except Exception as e:
    print('Issue communicating with the db')
    raise e
```

neverett commented on tasks/04_pages/load.py:
```diff
@@ -0,0 +1,28 @@
+from common import paths
```
Can this file be removed? Or did you want to keep it for loading sample HTML page data?

tgj505 replied:
missed this comment. removing it!

neverett commented on tasks/04_pages/post.py:
```diff
@@ -0,0 +1,46 @@
+from common import db_config, sql
+if __name__ == '__main__':
+    """
+    Cofirm the scraper has finished by ensuring all case_ids from the cases table
```
s/Cofirm/confirm

neverett commented on tasks/04_pages/post.py:
```diff
@@ -0,0 +1,46 @@
+    """
+
+    pages_query = 'SELECT count(case_id) FROM pages'
+    comparison_query = """
```
I think you can do a simpler query here to just return the difference as an integer instead of the full result of the join:
```
SELECT (SELECT count(*) FROM cases) - (SELECT count(*) FROM pages) AS row_count_result
```

neverett commented on tasks/04_pages/task.py:
```diff
@@ -0,0 +1,79 @@
+                num_cases = c.fetchone()[0]
+                c.execute(
+                    """
+                    DELETE FROM pages WHERE raw_text = '';
```
You probably want to execute this query before counting the number of pages left to scrape instead of after

tgj505 replied:
thanks! good call

neverett commented on tasks/04_pages/task.py:
```diff
@@ -0,0 +1,79 @@
+                    """
+                    )
+        except Exception as e:
+            print(f'Unable to count cases or clean empty rows: {e}')
```
For clarity, maybe this could be "Unable to count remaining case pages to scrape or clean empty rows"

neverett replied:
You probably also want to raise this exception instead of just printing it, since in the future we may have a top-level script that coordinates running all the separate tasks, and any exceptions should be bubbled up to its level to be handled there

tgj505 replied:
that makes sense!

neverett commented on tasks/04_pages/scraper.py:
```diff
@@ -0,0 +1,83 @@
+    finally:
+        c.close()
+        cnx.close()
+
```
You can delete two of these empty lines here

neverett commented on tasks/04_pages/scraper.py:
```diff
@@ -0,0 +1,83 @@
+
+def threaded_scraper(case):
+    """
+    scraper doesn't necessarily need to be run in threads
```
Maybe explain this a bit more -- is this to clarify that the name of the function doesn't necessarily mean it has to be run in threads? If so, maybe the function could be renamed?

tgj505 replied:
renamed and documented.

neverett commented on tasks/04_pages/scraper.py:
```diff
@@ -0,0 +1,83 @@
+
+
+
+def threaded_scraper(case):
```
Maybe add a type hint here to show that `case` should be a tuple.

tgj505 replied:
definitely!

neverett commented on tasks/04_pages/scraper.py:
```diff
@@ -0,0 +1,83 @@
+    except requests.exceptions.HTTPError as e:
+        print(f'Error: {e}, case: {case_number}')
+        logging.warning(f'{case_id}, {case_number}, fetch error: {e}')
+        return
```
You may want to remove this if you just want to log errors and move on. If not, then you'll want to raise the exception in addition to logging instead of returning.

tgj505 replied:
i think raising the exception makes sense here...

neverett commented on tasks/04_pages/scraper.py:
```diff
@@ -0,0 +1,83 @@
+            c.execute(query, (int(case_id), case_number, raw_text))        
+            
+    except Exception as e:
+        print(f'Error adding page to {db_config.pages} table: {e}')
```
You probably want to raise this exception in addition to printing, since if you're unable to write to the DB, you probably want the script to stop

neverett reviewed (commented):
This looks really polished to me, and really close to done! I left some comments on the code, and have one general point of feedback, which is that you may want to add a README to this directory to cover:
- how to run this task the first time
- a high-level summary of what happens when it's run
- roughly how long it takes (which will depend on people's individual setup, but you can give a rough estimate of round trip time per page and time to scrape HTML for the roughly 440k cases through the end of 2022)
- how to rerun the task if it stalls

If you'll be including any sample HTML page files that can be loaded with the `load.py` script, you may also want to cover that as well.

tgj505 committed:
Merge branch 'tom_scrape_pages' of https://github.com/datatoolsrc2023/nlrb_data into tom_scrape_pages

tgj505 committed:
refactoring for PR, README.md writing

tgj505 commented:
I've run through and incorporated fixes for all of Nikki's comments, including a `README.md` for the task.

tgj505 committed:
Delete load.py

remnant from an earlier data storage structure using .html files. obsolete.

tgj505 committed:
addressing PR comments

Refactored exceptions handling in 04_pages/task.py, clarified task restarting strategy (and in , and fixed other typos.

tgj505 committed:
refactoring post.py and Makefile

tgj505 committed:
refactoring post-scrape workflow

tgj505 committed:
task.py and post.py refactoring

Improving efficiency for the finishing aspects of the task.

neverett reviewed (approved):
Looks good to me!

tgj505 committed:
post.py minor formatting
